### PR TITLE
[SCM-994] Add credentials provider for JGit

### PR DESCRIPTION
**Pull request by kwin**

### CODE REVIEW

File: maven-scm-api/src/main/java/org/apache/maven/scm/provider/ScmProvider.java [hunk 1]
```diff
@@ -71,6 +71,15 @@ ScmProviderRepository makeProviderScmRepository( String scmSpecificUrl, char del
     ScmProviderRepository makeProviderScmRepository( File path )
         throws ScmRepositoryException, UnknownRepositoryStructure;
 
+    /**
+     * Sets the interactive mode. 
+     * @param isInteractive either {@code true} in case user may be prompted for information, otherwise {@code false}
+     * @since 2.0.0-M2
+     */
+    default void setInteractive( boolean isInteractive )
```
Comment (michael-o) [inline maven-scm-api/src/main/java/org/apache/maven/scm/provider/ScmProvider.java:79]: `boolean interactive`

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/pom.xml [hunk 2]
```diff
@@ -44,6 +44,17 @@
       <groupId>org.apache.maven.scm</groupId>
       <artifactId>maven-scm-provider-git-commons</artifactId>
     </dependency>
+    <dependency>
+      <artifactId>plexus-interactivity-api</artifactId>
+      <groupId>org.codehaus.plexus</groupId>
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/pom.xml:49]: swap groupId and artifactId

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/JGitScmProvider.java [hunk 3]
```diff
@@ -54,6 +58,21 @@
...
+
+    @Inject
+    public JGitScmProvider( Prompter prompter )
+    {
+        credentialsProvider = new PlexusInteractivityCredentialsProvider( prompter );
+        CredentialsProvider.setDefault( credentialsProvider );
+    }
+
+    @Override
+    public void setInteractive( boolean isInteractive )
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/JGitScmProvider.java:71]: `boolean interactive`

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.URIish;
+
+/**
+ * {@link CredentialsProvider} leveraging the {@link Prompter} component.
+ *
+ */
+public class PlexusInteractivityCredentialsProvider extends CredentialsProvider
+{
+    private boolean isInteractive;
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:38]: same here

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            catch ( PrompterException e )
+            {
+                throw new IllegalStateException( "Can not prompt user " + e.getMessage() , e );
+            }
+        }
+        return true;
+    }
+
+    public void setInteractive( boolean isInteractive )
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:89]: same here

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                else if ( item instanceof CredentialItem.InformationalMessage )
+                {
+                    prompter.showMessage( item.getPromptText() );
+                }
+                else
+                {
+                    throw new UnsupportedCredentialItem( uri, item.getClass().toString() );
+                }
+            }
+            catch ( PrompterException e )
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:81]: This cascade ignores the `interactive` flag.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java (hunk 6 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:81]: That is deliberate as JGit internally calls https://javadoc.io/static/org.eclipse.jgit/org.eclipse.jgit/5.13.1.202206130422-r/org/eclipse/jgit/transport/CredentialsProvider.html#isInteractive-- before calling get.

Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:81]: But Plexus Interactivity does not care. What do I miss here?

Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:81]: `get(...)` is not called if `isInteractive()` returns `false`. As Plexus Interactivity is only involved in `get(...)` it won't be called at all in that case.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/pom.xml (hunk 2 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/pom.xml:48]: Done in https://github.com/apache/maven-scm/pull/154/commits/617c43e02f33474dc9fc2c17ee400f874d18d789

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java (hunk 6 above)
Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:81]: Alright, I see. The provide is asked externally. Can you leave a simple comment for others not to trip over?

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java (hunk 4 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java]: Done in https://github.com/apache/maven-scm/pull/154/commits/617c43e02f33474dc9fc2c17ee400f874d18d789

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java (hunk 5 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java]: Done in https://github.com/apache/maven-scm/pull/154/commits/617c43e02f33474dc9fc2c17ee400f874d18d789

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/JGitScmProvider.java [hunk 7]
```diff
@@ -54,6 +58,21 @@
 public class JGitScmProvider
     extends AbstractGitScmProvider
 {
+    private final PlexusInteractivityCredentialsProvider credentialsProvider;
+
+    @Inject
+    public JGitScmProvider( Prompter prompter )
+    {
+        credentialsProvider = new PlexusInteractivityCredentialsProvider( prompter );
+        CredentialsProvider.setDefault( credentialsProvider );
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/JGitScmProvider.java:67]: yes. I think clashes are very unlikely, isn’t it?

[the remark this answers is not in the document]
Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/JGitScmProvider.java:67]: As long as we can assure that everything is a singleton and we don't have any race conditions. Setters like this expect all to be configured *before*  workers are spawned.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        for ( CredentialItem item : items )
+        {
+            try
+            {
+                get( uri, item );
+            }
+            catch ( PrompterException e )
+            {
+                throw new IllegalStateException( "Can not prompt user " + e.getMessage() , e );
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:58]: `"Cannot prompt user", e`

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            // works even in non-interactive mode
+            prompter.showMessage( item.getPromptText() );
+        }
+        else
+        {
+            if ( !interactive )
+            {
+                throw new UnsupportedCredentialItem( uri, 
+                        "Cannot provide " + item.getClass() + " in non-interactive mode" );
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java:76]: set the class name in single quotes.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java (hunk 9 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java]: Done in https://github.com/apache/maven-scm/pull/154/commits/521a09a78510a624271d8681c15e506c63b327c4.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java (hunk 8 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/PlexusInteractivityCredentialsProvider.java]: Done in https://github.com/apache/maven-scm/pull/154/commits/ac67c8ef4cdca949e0c27e703e7b2a8d96ecdb9c.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/JGitScmProvider.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/JGitScmProvider.java:67]: Yes, it is singleton. Therefore, I guess it is good.

**Comment by jira-importer** [comment]
Resolve #1220

**Pull request closed without merging**